proposals/data-classes.md

```diff
@@ -0,0 +1,372 @@
+    {
+        return $"{{{nameof(Username)} = {Username}, {nameof(Password)} = {Password}, {nameof(RememberMe)} = {RememberMe}}}";
+    }
+
+    public static bool operator ==(LoginResource resource1, LoginResource resource2)
+    {
+        return EqualityComparer<LoginResource>.Default.Equals(resource1, resource2);
+    }
+
+    public static bool operator !=(LoginResource resource1, LoginResource resource2)
+    {
+        return !(resource1 == resource2);
+    }
+}
+```
+
+### Equality
+
+First, the generation of equality support. Data members are only public
+fields and auto-properties. This allows data classes to have private
+implementation details without giving up simple equality semantics. There are
+a few places this could be problematic. For instance, only auto-properties
+are considered data members by default, but it's not uncommon to have some
+simple validation included in a property getter that does not meaningfully
+change the semantics, e.g.
+
+```C#
+{
+    ...
+    private int _field;
+    public int Field
+    {
+        get
+        {
+            Debug.Assert(_field >= 0);
+            return _field;
+        }
+        set { ... }
+    }
+}
+```
+
+To support these cases and provide an easy escape hatch, I propose a
+new attribute, `DataMemberAttribute` with a boolean flag argument on the
+constructor. This allows users to override the normal behavior and include
+or exclude extra members in equality. The previous example would now read:
+
+```C#
+{
+    ...
+    private int _field;
+
+    [DataMember(true)]
+    public int Field
+    {
+        get
+        {
+            Debug.Assert(_field >= 0);
+            return _field;
+        }
+    }
+}
+```
+
+Equality itself would be defined in terms of its data members. A `data` type
+is equal to another `data` type when there is an implicit conversion between
+the target type and the source type and each of the corresponding members
+are equal. The members are compared by `==` if it is available. Otherwise,
+the method `Equals` is tried according to overload resolution rules (st. an
+`Equals` method with an identity conversion to the target type is preferred
+over the virtual `Equals(object)` method).
+
+There is also one hidden data member, `EqualityContractOrigin`, that is
+always considered in equality. By default this member always returns the
+static type of its containing type, i.e. `typeof(Containing)`. This means
+that sub-classes are not, by default, considered equal to their base classes,
+or vice versa. This also ensures that equality is commutative and `GetHashCode`
+matches the results of `Equals`. These methods are virtual, so they can be
+overridden, but then it is the user's responsibility to ensure that they
+abide by the appropriate contract.
+
+`GetHashCode` would be implemented by calling `GetHashCode` on each of
+the data members and these would be combined using a symmetric operation
+(like `+`) to prevent ordering from affecting the results.
```

Comment:
not clear ot me why/how ordering could/would matter.  

Since you can't compare against a different type... your orders will be the same.

Reply:
You could certainly save the hash code and compare before/after reordering members.

Reply:
This is interesting because now you're providing an even stronger guarantee.  That the hashcode is not only safe across reordering.  But also safe across versioning of the compiler/language.  I'm ok with such a guarantee.  but we would really have to be specific about it.  

Personally though, i don't like it.   :)
 
I would much rather us get behind a *weaker* guarantee.  That only within a session could you even be sure that hte hashcodes would be the same.  I would even do what some systems do and introduce some random value so that each session produces new hashcodes.  That way, if you actually tried to take a dependency on hashcode values, that you would break almost immediately.

Reply:
@agocke [The documentation of `GetHashCode()`](https://docs.microsoft.com/en-us/dotnet/api/system.object.gethashcode) strongly warns against doing that.

Reply:
Basically, because collisions are far more likely.  frankly, i think from a language perspective we should say nothing more than: if you have the same member values in teh same session, you'll get teh same hashcode.    Then, from an impl perspective we should basically either call into HashCode.Combine (if it exists), or just emit the same GetHashCode code as we do for anonymous-types.

Reply:
Also: > is simpler

This isn't really a virtue (to me at least) for generated code.  We already have fairly good hashcodes for tuples/anonymous-types.   I would not want us to pick a poorer hashcode here (i.e. more likely to collide) and then regret that later on because customers are negatively impacted.

Reply:
Sure. You've convinced me. We should use whatever hash code combine we think is fast and provides a good distribution. Hash.Combine is probably a decent starting point.